### PR TITLE
env: Abstract DPDK dependences with SPDK env APIs.

### PR DESCRIPTION
**Pull request by johnmeneghini**
- Implemented a wrapper funtions for rte_config.h, rte_mempool.h, rte_lcore.h APIs.
    - Updated lib/env_dpdk with new spdk APIs to call rte functions
    - Changes to nvmf_tgt and iscsi_tgt app code; example and test code to call
      spdk env wrapper functions instead of rte_ functions directly.
    - Added various and sundry new definitions and macros for abstraction

### CODE REVIEW

**Comment by johnmeneghini** [comment]
I've updated this pull request with a second commit.   This second change breaks the main() function in the nvmf_tgt app into a separate file.  This allows consumers of the nvmf_tgt 
application to use the NVMF target infrastructure with their own initialization code rather than
forcing them call nvmf_tgt_main().  

spdk_app_opts_init(&opts);
/* fill in the opts struct */
spdk_nvmf_tgt_start(&opts);

This is a small change but it helped us.

File: include/spdk_internal/bdev.h [hunk 1]
```diff
@@ -181,7 +181,7 @@ spdk_bdev_io_from_ctx(void *ctx)
 }
 
 #define SPDK_BDEV_MODULE_REGISTER(init_fn, fini_fn, config_fn, ctx_size_fn)			\
-	static struct spdk_bdev_module_if init_fn ## _if = {					\
```
Comment (dverkamp-intel) [inline include/spdk_internal/bdev.h]: This seems unrelated; why is the static removed?

File: lib/event/app.c [hunk 2]
```diff
@@ -243,12 +241,14 @@ spdk_app_init(struct spdk_app_opts *opts)
 	char			*end;
 	struct spdk_env_opts env_opts = {};
 
+#ifndef SPDK_NO_RLIMIT
```
Comment (dverkamp-intel) [inline lib/event/app.c]: The RLIMIT change should be in a separate patch.

File: lib/event/reactor.c [hunk 3]
```diff
@@ -1,43 +1,45 @@
...
- *   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
- *   OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
- *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
- *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
- *   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
- *   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
- *   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
- *   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
- */
+*   BSD LICENSE
```
Comment (dverkamp-intel) [inline lib/event/reactor.c]: Please drop these unrelated whitespace changes in the copyright header.

File: lib/json/json_write.c [hunk 4]
```diff
@@ -147,7 +147,7 @@ emit_buf_full(struct spdk_json_write_ctx *w, const void *data, size_t size)
 	}
 
 	/* Recurse to emit the rest of the data. */
-	return emit(w, data + buf_remain, size - buf_remain);
+	return emit(w, (void *)((int *)data + buf_remain), size - buf_remain);
```
Comment (dverkamp-intel) [inline lib/json/json_write.c]: This seems unrelated and should be split to a different commit.

I'm also not sure what it is trying to fix. The (int *) cast will result in incorrect pointer math; we want to add buf_remain in bytes, not ints.

File: lib/nvmf/rdma.c [hunk 5]
```diff
@@ -52,8 +52,6 @@
 #include "spdk/nvmf_spec.h"
 #include "spdk/string.h"
 #include "spdk/trace.h"
-#include "spdk/util.h"
```
Comment (dverkamp-intel) [inline lib/nvmf/rdma.c]: Seems unrelated, and this file uses definitions from util.h (e.g. spdk_min), so it shouldn't be removed.

File: lib/vhost/rte_vhost/virtio_net.c [hunk 6]
```diff
@@ -95,7 +97,7 @@ static inline void __attribute__((always_inline))
 do_flush_shadow_used_ring(struct virtio_net *dev, struct vhost_virtqueue *vq,
 			  uint16_t to, uint16_t from, uint16_t size)
 {
-	rte_memcpy(&vq->used->ring[to],
```
Comment (dverkamp-intel) [inline lib/vhost/rte_vhost/virtio_net.c]: The files in rte_vhost are a temporary fork of the DPDK rte_vhost library, and we should not modify them if at all possible to make updates simpler. Also, the vhost code is inextricably tied to DPDK, so there's no reason to remove rte_ calls from it.

File: mk/spdk.common.mk [hunk 7]
```diff
@@ -48,6 +48,7 @@ OS := $(shell uname)
 COMMON_CFLAGS = -g $(C_OPT) -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wmissing-declarations -fno-strict-aliasing -march=native -m64 -I$(SPDK_ROOT_DIR)/include
 
 COMMON_CFLAGS += -include $(SPDK_ROOT_DIR)/config.h
+COMMON_CFLAGS += -I$(SPDK_ROOT_DIR)/lib/env_dpdk
```
Comment (dverkamp-intel) [inline mk/spdk.common.mk]: I don't think this should be necessary; nothing should be including files from lib/env_dpdk directly, since all of the public API should be declared in include/spdk/env.h.

**Comment by dverkamp-intel** [comment]
I merged the nvmf_tgt main() split as commit b512e01a50d44795aa3ee89c8d25a0acc6e5cf96.

I also pulled out the removal of rte_* includes from files that don't use DPDK functions at all as 1e4920b2614f06bab70d848f080720becc3800eb.

It would still be good to split up the rest of the change into more manageable pieces.

**Comment by johnmeneghini** [comment]
Thanks Daniel.   I’ll address your comments today.  I was busy yesterday.

John Meneghini
john.a.meneghini@gmail.com

File: include/spdk_internal/bdev.h (hunk 1 above)
Comment (johnmeneghini) [inline include/spdk_internal/bdev.h]: With change 8ea1349 I can remove these.

File: lib/event/app.c (hunk 2 above)
Comment (johnmeneghini) [inline lib/event/app.c]: I'll move this to a separate patch.

File: lib/event/reactor.c (hunk 3 above)
Comment (johnmeneghini) [inline lib/event/reactor.c]: I thought the check-format.sh script would take care of this.  I'll run this script again, but this is the second time running check-format.sh seems to have produced the wrong outcome for me.

File: lib/json/json_write.c (hunk 4 above)
Comment (johnmeneghini) [inline lib/json/json_write.c]: I'll look at this.

File: lib/nvmf/rdma.c (hunk 5 above)
Comment (johnmeneghini) [inline lib/nvmf/rdma.c]: I'll look at it.

File: lib/vhost/rte_vhost/virtio_net.c (hunk 6 above)
Comment (johnmeneghini) [inline lib/vhost/rte_vhost/virtio_net.c]: OK. I'll revert all of the changes to virtio_net.c

File: mk/spdk.common.mk (hunk 7 above)
Comment (johnmeneghini) [inline mk/spdk.common.mk]: Agreed.  I'll look at this.

**Comment by johnmeneghini** [comment]
Hi Daniel.

While reworking this pull request I noticed that Ben’s commit 16ae587 <https://github.com/spdk/spdk/commit/16ae5879664573002db6379f5537a8640264100a> has made some conflicting rte_lcore changes that conflict with my lcore changes.

Do you guys have any other changes to env.h or env.c ?

I’m going to refactor this pull request and send it again.   I’d prefer to make this one big commit, but I can break this into 3: once for mempool, lcore, and ring.

Thoughts?

John Meneghini
john.a.meneghini@gmail.com

**Comment by johnmeneghini** [comment]
I've created a new pull request for these changes so I'm closing this one.
